### PR TITLE
[JENKINS-47531] Allow separate Git LFS credentials

### PR DESCRIPTION
**Pull request by mephi42**

### CODE REVIEW

**Comment by mephi42** [comment]
The build keeps failing on different builders due to timeouts when downloading artifacts. Is there a "retrigger" button somewhere?

**Comment by MarkEWaite** [comment]
@mephi42 the only retrigger is to close and reopen the pull request.  The infra team is working on the problem that is causing unreliable builds on ci.jenkins.io.  I will evaluate the pull request in my infrastructure as well before merging the change.

**Comment by mephi42** [comment]
I have fixed unit tests (I was creating GitLFSPullTrait incorrectly) and simplified getting the Job object.

**Comment by mephi42** [comment]
@MarkEWaite @stephenc, are there any changes I need to make in order to get this in?

**Comment by MarkEWaite** [comment]
I need to evaluate the pull request in more detail, then merge it.  If the required git client plugin has not yet released with the needed API additions, that will also need to happen before this is merged and released.

**Comment by MarkEWaite** [comment]
I needed to integrate a more version of the git client library into the git plugin as part of #558 .  I integrated this change into that same pull request as an easier way to update the git plugin to use the meet the new credential requirements from the git client plugin.

If you can help evaluate #558 that will also help complete the evaluation of this pull request.

**Comment by MarkEWaite** [comment]
@stephenc mentioned in a [comment](https://issues.jenkins-ci.org/browse/JENKINS-47531?focusedCommentId=318027&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-318027) on [JENKINS-47531](https://issues.jenkins-ci.org/browse/JENKINS-47531) that he had additional comments that would be better suited to a pull request.  

Have those comments been received?

Have they been addressed?

**Comment by mephi42** [comment]
@MarkEWaite, no, unfortunately I have not received @stephenc's comments yet.

**Comment by samrocketman** [comment]
@bienstock did you also test a plugin upgrade for freestyle and pipeline? I.e. upgrade existing stable plugin with configs to this snapshot?

**Comment by bienstock** [comment]
@samrocketman These were the test specs:
``` 
Jenkins 2.107.3 
git plugin version 4.0.0-rc1679.0ee1ac328cfd(the PR version) 
git-client version 3.0.0-beta3
```
As per [JENKINS-47531](https://issues.jenkins-ci.org/browse/JENKINS-47531?focusedCommentId=344116&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-344116), you can't install the PR on the latest stable version, only on the experimental ```4.0.0-beta3``` version.

**Comment by jmcfallargo** [comment]
LGTM?

File: src/main/resources/hudson/plugins/git/extensions/impl/GitLFSPull/config.groovy [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package hudson.plugins.git.extensions.impl.GitLFSPull;
```
Comment (darxriggs) [inline src/main/resources/hudson/plugins/git/extensions/impl/GitLFSPull/config.groovy]: Semicolons are optional in Groovy and normally are not used.
Therefore I suggest to remove it.
(Yes, there many examples in the code that aren't good examples.)

File: src/main/java/hudson/plugins/git/extensions/impl/GitLFSPull.java [hunk 2]
```diff
@@ -76,6 +96,13 @@ public String toString() {
 
     @Extension
     public static class DescriptorImpl extends GitSCMExtensionDescriptor {
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project,
+                                                     @QueryParameter String url,
+                                                     @QueryParameter String credentialsId) {
+            return new UserRemoteConfig.DescriptorImpl().doFillCredentialsIdItems(
```
Comment (darxriggs) [inline src/main/java/hudson/plugins/git/extensions/impl/GitLFSPull.java]: I am not sure if this intended to be reused like this. @MarkEWaite what's your opinion?

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 3]
```diff
@@ -852,6 +840,26 @@ public GitClient createClient(TaskListener listener, EnvVars environment, Run<?,
...
+                job instanceof Queue.Task
+                        ? Tasks.getDefaultAuthenticationOf((Queue.Task)job)
+                        : ACL.SYSTEM,
+                URIRequirementBuilder.fromUri(url).build()
+        );
+        CredentialsMatcher ucMatcher = CredentialsMatchers.withId(credentialsId);
+        CredentialsMatcher idMatcher = CredentialsMatchers.allOf(ucMatcher, GitClient.CREDENTIALS_MATCHER);
+        StandardUsernameCredentials credentials = CredentialsMatchers.firstOrNull(urlCredentials, idMatcher);
+        if (credentials != null) {
+            if (job != null && job.getLastBuild() != null) {
```
Comment (darxriggs) [inline src/main/java/hudson/plugins/git/GitSCM.java]: Now this could also be merged with the previous `if` statement.

File: src/main/java/hudson/plugins/git/GitSCM.java [hunk 4]
```diff
@@ -852,6 +840,26 @@ public GitClient createClient(TaskListener listener, EnvVars environment, Run<?,
         return c;
     }
 
+    public static StandardCredentials resolveCredentials(Job job, String credentialsId, String url) {
```
Comment (darxriggs) [inline src/main/java/hudson/plugins/git/GitSCM.java]: AFAIK that's only extracted the functionality into a method without a functional change - right?
I suggest to extract this change into a dedicated commit to separate functional changes from refactorings.

Comment (darxriggs) [inline src/main/java/hudson/plugins/git/GitSCM.java]: I suggest to reduce the scope to `private`. Otherwise this method gets part of the public API and has to be maintained like this. Or was this intended?

**Comment by darxriggs** [comment]
@mephi42 / @MarkEWaite can you please reopen this one?

**Comment by darxriggs** [comment]
@mephi42 can you please resolve the merge conflict and work on the review comments or provide an estimated time when you can do so?

**Comment by mephi42** [comment]
@darxriggs I'm sorry, I don't work on the project which had the relevant Jenkins setup anymore. So at the moment I don't have plans to work on this PR. FWIW I won't mind if someone picks it up though.
